docs(tutorials/jokes): update to be compatible with `v2_dev` (#7162)
remix-run/remix: docs/tutorials/jokes.md

So we've got a bit of extra work to do to avoid this development time problem.

> Note that this isn't a remix-only problem. Any time you have "live reload" of server code, you're going to have to either disconnect and reconnect to databases (which can be slow) or do the workaround [that's in our docs][singleton-workaround-docs].

Reviewer: ```suggestion
Note that this isn't a remix-only problem. Any time you have "live reload" of server code, you're going to have to either disconnect and reconnect to databases (which can be slow) or [use the `global` singleton workaround][singleton-workaround-docs].
```

💿 Copy this into two new files called `app/utils/singleton.server.ts` & `app/utils/db.server.ts`

```ts filename=app/utils/singleton.server.ts
export const singleton = <Value>(
  name: string,
  valueFactory: () => Value
): Value => {
  const g = global as any;
  g.__singletons ??= {};
  g.__singletons[name] ??= valueFactory();
  return g.__singletons[name];
};
```

```ts filename=app/utils/db.server.ts
import { PrismaClient } from "@prisma/client";

import { singleton } from "./singleton.server";

// Hard-code a unique key, so we can look up the client when this module gets re-imported
const db = singleton("prisma", () => new PrismaClient());

> db.$connect();

Reviewer 2: Put the connect call within the callback, otherwise you'll call `$connect` on every file change which is probably not good.

Or just remove the connect call altogether. It's probably not necessary anyway.

Note that this isn't a remix-only problem. Any time you have "live reload" of server code, you're going to have to either disconnect and reconnect to databases (which can be slow) or do the workaround I'm about to show you.

💿 Copy this into two new files called `app/utils/singleton.server.ts` & `app/utils/db.server.ts`

```ts filename=app/utils/singleton.server.ts
// Since the dev server re-requires the bundle, do some shenanigans to make
// certain things persist across that 😆
// Borrowed/modified from https://github.com/jenseng/abuse-the-platform/blob/2993a7e846c95ace693ce61626fa072174c8d9c7/app/utils/singleton.ts

export const singleton = <Value>(
  name: string,
  valueFactory: () => Value
): Value => {
  const g = global as any;
  g.__singletons ??= {};
  g.__singletons[name] ??= valueFactory();
  return g.__singletons[name];
};

> ```

Reviewer: We now have [docs for how to do this](https://remix.run/docs/en/main/guides/manual-mode#keeping-in-memory-server-state-across-rebuilds), so could you update this code to be consistent with that and maybe link to those docs instead?

Author: This is the same as we have in `indie-stack`, where I just took over whatever @jenseng had in his code + adapted it to accept a function (just like Kent did in his personal website and the epic stack)

I would argue to bring the docs more in line with what we have in there tbh

Reviewer: Ok, let's update the docs to use `singleton` as defined here, but then link to those docs here since the docs will have a more detailed explanation.
